Add plumbing for tracing

igorbernstein2 opened the pull request:
This has been extracted from #613

This doesn't provide any extra functionality. Instead it just sets up the foundation for integrating tracing in to gax.
The general idea is to add a tracing abstraction layer that all parts of gax can call into. This is accomplished by introducing
- ApiTracer interface that has methods for all annotations that a span should contain
- ApiTracerFactory interface: since an ApiTracer is stateful and has a 1:1 mapping with operations, a factory is introduced to cleanly switch implementations of the tracer

By default a NoopApiTracerFactory is configured that will return NoopApiTracers. In the future a parallel OpenCensusApiTracerFactory will be added. A client implementation will be able to opt into tracing by setting the factory in StubSettings. The actual ApiTracer will be propagated by CallContext.

vam-google commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonCallContext.java:
```diff
@@ -230,11 +242,30 @@ public Credentials getCredentials() {
+  @Override
+  public ApiTracer getTracer() {
+    if (tracer == null) {
+      return NoopApiTracer.create();
```
Please check the comment for `TRACER_KEY`.

igorbernstein2 replied:
ack

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/NoopApiTracer.java:
```diff
@@ -0,0 +1,118 @@
+
+  private NoopApiTracer() {}
+
+  public static ApiTracer create() {
```
This is a create method which does not create. Please change it to something like `getInstance()` or `getDefault()` or anything else as long as it indicates that there is no creation and two calls to this method will return same object.

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/ApiTracer.java:
```diff
@@ -0,0 +1,112 @@
+   * Signals that the overall operation has failed and no further attempts will be made. The tracer
+   * is now considered closed and should no longer be used.
+   */
+  void operationFailed(Throwable error);
```
Can you please add the documentation for arguments here and in other places? We do not add `@return` documentation element but usually add the `@param` (in some existing code it is missing, but that is not a pattern we should try to follow in new code).

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/ApiTracer.java:
```diff
@@ -0,0 +1,112 @@
+   * {@link Scope} removes any context that the underlying implementation might've set in {@link
+   * #inScope()}.
+   */
+  interface Scope extends AutoCloseable {
```
I don't remember if we've already discussed this, but why do we want to add an abstraction on top of opencensus here? Why not just expose it as is? Opencensus itself already has a lot of power and flexibility. Specifically:
- opencensus is already good in `noop` world (i.e. just calling it and not using the reports is effectively a resource efficient noop tracer already).
- opencensus already has pluggable modules (so we can specify which "reporting system to use" (or whatever that is called properly)) and is flexible to the extent that maybe we do not even need to be able to plug a different reporting mechanism (not opencensus). Exposing opencensus directly can be an option, and this allows us to not have another level of abstraction (like this interface, for example)

igorbernstein2 replied:
There are couple of reasons for this approach:
1. I didn't want to sprinkle arbitrary annotation strings throughout the code base. I think it's a lot easier to reason about if all of the annotations are centralized in a single package.
2. I would like to include a fair amount of attributes with the annotations like response counts, these numbers are only needed for tracing and I don't want to pollute business logic at call sites with the accounting. Also some of the attributes cut across layers: like an operation span having an attempt count
3. Finally, an unintentional benefit is that hiding the surface of opencensus gives us a way out in case we get into dependency hell like guava

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/ApiTracerFactory.java:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2017 Google LLC
```
Please correct year here and in other new files of this PR. At least this one is definitely wrong. As for 2018 vs 2019, we can keep the year of creation (i.e. 2018), though I feel the year when it gets into master is the proper one (i.e. 2019)... Also it kind of gives the privilege to submit first 2019 headers in the repo =).

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/NoopApiTracerFactory.java:
```diff
@@ -0,0 +1,56 @@
+import com.google.api.core.InternalApi;
+
+/**
+ * Factory that will build fake {@link ApiTracer}s.
```
Lets not call noop tracer "fake" tracer. Noop is still a valid tracer, it is not fake.

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/NoopApiTracerFactory.java:
```diff
@@ -0,0 +1,56 @@
+
+  @Override
+  public boolean equals(Object obj) {
+    return obj instanceof NoopApiTracerFactory;
```
Both `equals()` and `hashCode()` look odd. I feel the proper way to achieve "same result" (all `NoopApiTracerFactory` are equal) is to have private constructor here and getInstance static method (i.e. converting this to singleton). 

Also usage of `instanceof` makes it look even "odder", since rigorous implementations of equals is supposed to compare actual classes (`getClass() == obj.getClass()`) instead of `instanceof` (yes, `instanceof` is used sometimes, for example in Hibernate entities `equals()` implementations, because in real life scenarios we would have to compare instance with proxy instance, but it is a special case). It is a final class, so `instanceof` is effectively same as comparing the actual class... Well what I mean by all this mutter is that the two methods look odd from all angles, and I would be nice to not have them at all.

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/SpanName.java:
```diff
@@ -0,0 +1,66 @@
+  /**
+   * Creates a new instance of the name.
+   *
+   * @param clientName The name of the client. ie BigtableData
```
Using outer scope examples (BigtableData) can be confusing for the users (they may start searching for BigtableData client in gax's codebase). I understand that examples are usually good for understanding, but lets keep gax independent from big table on all levels (including documentation).

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/tracing/SpanName.java:
```diff
@@ -0,0 +1,66 @@
+  public abstract String getMethodName();
+
+  /** Creates a new instance with the clientName overriden. */
+  public SpanName withClientName(String clientName) {
```
Why would we need to override client name? It seems like an impossible scenario. If there is no real need to have it, please remove the method (or clarify what exactly "client" means here)

igorbernstein2 replied:
This is needed for manual wrappers like bigtable. I need to disambiguate between traces created by the gapic layer and the manually written callable chain. The outer bigtable traces will be called something like `BigtableData.ReadRows` while the underlying gapic generated child traces will be named `BaseBigtableData.ReadRows`

vam-google replied:
But in that case is it still "client". Is'nt it "caller", "layer" or something? My understanding is that there is one client: bigtable client. This is what a user expects: he has a client (bigtable) and that is it. I'm find with keeping this as long as it is clear what "client" is here (it can be done without serivce-specific names, just refering to them as "GAPIC (generated) client" and "manual wrapper client around the generated one".

vam-google replied:
But in that case is it still "client". Is'nt it "caller", "layer" or something? My understanding is that there is one client: bigtable client. This is what a user expects: he has a client (bigtable) and that is it. I'm fine with keeping this as long as it is clear what "client" is here (it can be done without serivce-specific names, just refering to them as "GAPIC (generated) client" and "manual wrapper client around the generated one".

igorbernstein2 commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java:
```diff
@@ -60,6 +64,8 @@
 @BetaApi("Reference ApiCallContext instead - this class is likely to experience breaking changes")
 @InternalExtensionOnly
 public final class GrpcCallContext implements ApiCallContext {
+  private static final CallOptions.Key<ApiTracer> TRACER_KEY = Key.create("gax.tracer");
```
Unfortunately that won't play nice with `merge(ApiCallContext)`. I need to know if someone explicitly set a `NoopApiTracer` in which case it should override the current one or if its the default one then I should keep the existing one. `Key.createWithDefault` prevents me from distinguishing between the 2 scenarios. 

I added a test case to make sure this behavior is respected

vam-google replied:
I do not understand. Why does it matter? `NoopApiTracer` is a singleton anyways, so there can be only one `NoopApiTracer` instance in the application, and both returned as default or explicitly set will be exactly same thing. Why does it all matter? Even if it does, then maybe just check that the "new tracer" is same as default one and simply to not set it, as we already happy with what we have (i.e. modify `withTracers()`).

igorbernstein2 replied:
It matters when a Tracer is set on both an ApiCallContext and a ClientContext. A ClientContext is supposed to provide the default callcontext with a default tracer. When a non-noop tracer is set on the ClientContext and the caller explicitly sets a noop tracer for the call, then the least surprising behavior is for the call chain to use the noop tracer.

vam-google commented on gax-httpjson/src/test/java/com/google/api/gax/httpjson/HttpJsonCallContextTest.java:
```diff
@@ -152,4 +153,26 @@ public void testMergeWithTimeout() {
+    HttpJsonCallContext ctxWithDefaultTracer = HttpJsonCallContext.createDefault();
+    ApiTracer defaultTracer = ctxWithDefaultTracer.getTracer();
+
+    // Explicit tracers override default tracers
```
Does "default tracers" (as plural) make sense? I thought it is explicitly one (per context class type, but in this context I think it is not important).

vam-google commented on gax-httpjson/src/test/java/com/google/api/gax/httpjson/HttpJsonCallContextTest.java:
```diff
@@ -152,4 +153,26 @@ public void testMergeWithTimeout() {
+    Truth.assertThat(ctxWithDefaultTracer.merge(ctxWithExplicitTracer).getTracer())
+        .isSameAs(explicitTracer);
+
+    // Default tracer do not override explicit tracers
```
Typo, please use either "tracers" or "does" (probably "tracers" because they override  "explicit trcerS").

igorbernstein2 commented:
I think addressed all of the concerns. Thanks for the review! I'm gonna go ahead and merge this. If there is anything else I should change I'll do it in a follow up PR
